### PR TITLE
New systemd::timer define type

### PR DESCRIPTION
**Pull request by traylenator**
The defined type allows easy creation of the timer and service unit at the same time.

Examples of usage

```puppet
systemd::timer{'runoften.timer':
  timer_source   => "puppet:///modules/${module_name}/runoften.timer",
  service_source => "puppet:///modules/${module_name}/runoften.service",
}
```
or a trivial daily run as:

```puppet
systemd::timer{'daily.timer':
  timer_content  => "[Timer]\nOnCalendar=daily\nRandomizedDelaySec=1d\n",
  service_content => "[Service]\nType=oneshot\nExecStart=/usr/bin/touch /tmp/file",
}
```

If neither `service_content` or `service_source` are specified then no service unit will be created.

The service unit name can also be specified.

```puppet
systemd::timer{'daily.timer':
  timer_content  => "[Timer]\nOnCalendar=daily\nRandomizedDelaySec=1d\nUnit=touch-me-today.service",
  service_unit   => 'touch-me-today.service',
  service_content => "[Service]\nType=oneshot\nExecStart=/usr/bin/touch /tmp/file",
}
```

Fixes: #118

### CODE REVIEW

File: manifests/timer.pp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+define systemd::timer (
+  Enum['present', 'absent', 'file']        $ensure    = 'present',
+  Stdlib::Absolutepath                     $path      = '/etc/systemd/system',
+  Optional[String]                         $timer_content   = undef,
+  Optional[String]                         $timer_source    = undef,
+  Optional[String]                         $service_content   = undef,
+  Optional[String]                         $service_source    = undef,
+  String                                   $owner     = 'root',
+  String                                   $group     = 'root',
+  String                                   $mode      = '0444',
```
Comment (bastelfreak) [inline manifests/timer.pp]: can you use https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/REFERENCE.md#stdlibfilemode ?

File: manifests/timer.pp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+#
+# @param service_unit
+#   If set then the service_unit will have this name.
+#   If not set the service unit has the same name
+#   as the timer unit with s/.timer/.service/
+#
+define systemd::timer (
+  Enum['present', 'absent', 'file']        $ensure    = 'present',
+  Stdlib::Absolutepath                     $path      = '/etc/systemd/system',
+  Optional[String]                         $timer_content   = undef,
```
Comment (bastelfreak) [inline manifests/timer.pp]: for all this strings, an empty string should be prohibited. Can you update them to use `String[1]`?

File: manifests/timer.pp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+  if $service_unit {
+    $_service_unit = $service_unit
+  } else {
+    $_service_unit = "${basename($name,'.timer')}.service"
+  }
+
+  contain systemd::systemctl::daemon_reload
+
+  if $service_content != undef or $service_source != undef {
```
Comment (bastelfreak) [inline manifests/timer.pp]: if you want, you can shorten this to:
```suggestion
  if $service_content or $service_source {
```

File: README.md [hunk 4]
```diff
@@ -129,6 +129,38 @@ file { '/etc/tmpfiles.d/foo.conf':
 ~> Class['systemd::tmpfiles']
 ```
 
+### timer units
+Create a timer unit and a service unit to execute from
+that timer at the same time.
+
+```puppet
+systemd::timer{'runoften.timer':
```
Comment (ekohl) [inline README.md]: Should the timer be automatically added? It's a bit redundant.

Comment (traylenator) [reply README.md]: @ekohl I'm not sure what you mean here?

Comment (ekohl) [reply README.md]: Should it accept `systemd::timer{'runoften':` and in timer.pp add `.timer` internally?

Comment (traylenator) [reply README.md]: Got you, that `.timer` I went through this myself. I thought it was more consistent with `systemd::unit_file` for instance.

Certainly it would remove all that basename dancing.

Comment (traylenator) [reply README.md]: I think I prefer it with the .timer forced, it makes it consistent with
the `service_unit` parameter as well.

**Comment by traylenator** [comment]
Of course there is another unrelated exercise to write a systemd provider type for the existing cron type.

File: manifests/timer.pp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    systemd::unit_file{$_service_unit:
+      ensure    => $ensure,
+      content   => $service_content,
+      source    => $service_source,
+      path      => $path,
+      owner     => $owner,
+      group     => $group,
+      mode      => $mode,
+      show_diff => $show_diff,
+      before    => Systemd::Unit_file[$name],
```
Comment (ekohl) [inline manifests/timer.pp]: When it's absent, shouldn't this be the other way around?

File: manifests/timer.pp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+) {
+  assert_type(Pattern['^.+\.timer$'],$name)
+
+  if $service_unit {
+    $_service_unit = $service_unit
+  } else {
+    $_service_unit = "${basename($name,'.timer')}.service"
+  }
+
+  contain systemd::systemctl::daemon_reload
```
Comment (ekohl) [inline manifests/timer.pp]: Doesn't the unit file handle this already?

File: manifests/timer.pp (hunk 5 above)
Comment (traylenator) [reply manifests/timer.pp]: Technically yes , when enable it's  important since a timer will fail to start if the service does not exist.

It should be  for enable

* create unit files
* run daemon-reload
* enable/start units

as for disable

* stop/disable units 
* remove files
* daemon-reload

so in fact the relation ship betwee the creation of the unit files is irreverent and this can go completely in fact.

File: manifests/timer.pp (hunk 6 above)
Comment (traylenator) [reply manifests/timer.pp]: Not quite in my experience the order is not correct and deamon-reload often happens after the service is 
managed, looking a bit harder now. `unit_file` contains:

```puppet
   include systemd    # which contains  daemon_reload'
     file { "${path}/${name}":
     ensure    => $_ensure,
     show_diff => $show_diff,
     notify    => Class['systemd::systemctl::daemon_reload'],
  }
service { $name:
  ensure    => $active,
  subscribe => File["${path}/${name}"],
  require   => Class['systemd::systemctl::daemon_reload'],
}

```
there is nothing in there that forces the exec of daemon-reload to happen before the service is managed.

Probably if unit_file itself contained `systemd::systemctl::daemon_reload`  or prossibly `contain systemd` `it might work.

Comment (ekohl) [reply manifests/timer.pp]: > there is nothing in there that forces the exec of daemon-reload to happen before the service is managed.

The `service` has a `require => Class['systemd::systemctl::daemon_reload']`. Doesn't that take care of it?

Comment (traylenator) [reply manifests/timer.pp]: No it does not. 

The class needs to be contained where the require is done. 
With include class you are saying I require that class be included before.
With contain class you are saying I require that class be included and do everything inside that class.

`require -> Class['abc']`  on its own is almost always not what you want with out a contain (or anchor)

Comment (traylenator) [reply manifests/timer.pp]: Another MR s/include/contain/ problably thing to do. Can do in a week or two ....

Comment (ekohl) [reply manifests/timer.pp]: That's not my understanding of how contain works. AFAIK `contain x` in `class a` it means that if I `require => Class['a']` I'll be sure that `class x` is executed before. However, the service still requires it and within the catalog it should be ordered correctly.

That said, with `include` and just `notify => Class['daemon_reload']` you can't be sure with just `require => Systemd::Unit_file['...']` that it has reloaded.

Comment (alexjfisher) [inline manifests/timer.pp]: I think the `contain` here is likely to cause circular dependencies when you have eg two systemd::timer resources in different profiles and you try to also have an order between those profiles.

**Comment by mmoll** [comment]
what's the status here?

**Comment by bastelfreak** [comment]
@traylenator can you take a look at the above conversations?

File: manifests/timer.pp (hunk 6 above)
Comment (traylenator) [reply manifests/timer.pp]: contain is gone.  relying on the include inside unit_file.

**Comment by traylenator** [comment]
I believe all comments are addressed.

**Comment by alexjfisher** [comment]
Regarding `systemd::systemctl::daemon_reload` I think (in a separate PR), the docs need tweaking a bit.
in unit_file.pp a [relationship](https://github.com/camptocamp/puppet-systemd/blob/f81339ad10516905e460d2899432ce92d83a35d2/manifests/unit_file.pp#L102) is created, but *only* when [enable or active](https://github.com/camptocamp/puppet-systemd/blob/f81339ad10516905e460d2899432ce92d83a35d2/manifests/unit_file.pp#L89) are used.  If you are managing the service resource yourself, you have to create this relationship yourself.

Also `systemd::systemctl::daemon_reload` is only relevant for puppet 5.  Since puppet 6.1.0, the systemd service provider in core puppet handles the reload itself.
Also see https://github.com/voxpupuli/puppet-prometheus/pull/390/files#diff-0ac866172efd29d0ddc2aa75ae6f182dR188-R192 for an example.

File: manifests/timer.pp [hunk 7]
```diff
@@ -0,0 +1,99 @@
+# Create a timer and optionally a service unit
+# to execute with the timer unit.
+#
+# @api public
+#
+# @see systemd.timer
```
Comment (alexjfisher) [inline manifests/timer.pp]: Was this just a placeholder? Is this supposed to link to an external website?

File: manifests/timer.pp [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+  Optional[String[1]]                      $timer_content = undef,
+  Optional[String[1]]                      $timer_source = undef,
+  Optional[String[1]]                      $service_content = undef,
+  Optional[String[1]]                      $service_source  = undef,
+  String[1]                                $owner = 'root',
+  String[1]                                $group = 'root',
+  Stdlib::Filemode                         $mode = '0444',
+  Optional[Systemd::Unit]                  $service_unit = undef,
+  Boolean                                  $show_diff = true,
+  Optional[Variant[Boolean, Enum['mask']]] $enable = undef,
```
Comment (alexjfisher) [inline manifests/timer.pp]: `enable` and `active` aren't documented above.  If either are set a service resource will be [created](https://github.com/camptocamp/puppet-systemd/blob/f81339ad10516905e460d2899432ce92d83a35d2/manifests/unit_file.pp#L90).  Is this correct?

File: manifests/timer.pp (hunk 7 above)
Comment (traylenator) [reply manifests/timer.pp]: I copied it from the other types:

```
tmpfile.pp:# @see systemd-tmpfiles(8)
tmpfiles.pp:# @see systemd-tmpfiles(8)
unit_file.pp:# @see systemd.unit(5)
```

I added the (5) now.

Now if they she be there at all is another matter.

**Comment by traylenator** [comment]
I improved the examples to document both manual and and automatic cases for managing the service.

**Comment by traylenator** [comment]
For the puppet > 6.1.4 case I think this is just a documentation fix which I have done.
I am not including the `systemd::systemctl::daemon_reload` anywhere anymore in this file at least.

File: manifests/timer.pp (hunk 7 above)
Comment (traylenator) [reply manifests/timer.pp]: @alexjfisher  remove it?

File: manifests/timer.pp (hunk 8 above)
Comment (alexjfisher) [reply manifests/timer.pp]: Looks like this works.  Slight concern that this is by accident and the `service` resource type wasn't intended to manage timers.
For instance, `puppet resource service` doesn't return any timers.  See the provider's [`instance`](https://github.com/puppetlabs/puppet/blob/4fcf3d953a0467555fc5477a2e28b1d5fd6819ec/lib/puppet/provider/service/systemd.rb#L32) method.

Comment (alexjfisher) [reply manifests/timer.pp]: Less concerned now that I've read [this](https://github.com/puppetlabs/puppet/blob/4fcf3d953a0467555fc5477a2e28b1d5fd6819ec/lib/puppet/provider/service/systemd.rb#L9).

File: manifests/timer.pp (hunk 7 above)
Comment (alexjfisher) [reply manifests/timer.pp]: Or link to eg. https://www.freedesktop.org/software/systemd/man/systemd.timer.html ??

File: README.md [hunk 9]
```diff
@@ -129,6 +129,80 @@ file { '/etc/tmpfiles.d/foo.conf':
...
+RandomizedDelaySec=1d
+EOT
+
+$_service = @(EOT)
+[Service]
+Type=oneshot
+ExecStart=/usr/bin/touch /tmp/file
+EOT
+
+if versioncmp($facts['puppetversion'],'6.1.0') < 0 {
```
Comment (ekohl) [inline README.md]: When using `enable` and/or `active` I think that it already takes care of this, right? Might be good to explain this to the user.

File: README.md [hunk 10]
```diff
@@ -129,6 +129,80 @@ file { '/etc/tmpfiles.d/foo.conf':
...
+  include systemd::systemctl::daemon_reload
```
Comment (alexjfisher) [inline README.md]: This will always end up included by the main class (which is included by unit_file).
https://github.com/camptocamp/puppet-systemd/blob/f81339ad10516905e460d2899432ce92d83a35d2/manifests/init.pp#L113

File: README.md [hunk 11]
```diff
@@ -129,6 +129,80 @@ file { '/etc/tmpfiles.d/foo.conf':
...
+  Systemd::Timer['daily.timer'] ~> Class[Systemd::Systemctl::Daemon_reload]
```
Comment (alexjfisher) [inline README.md]: This relationship already exists??
Should be
```
Class[Systemd::Systemctl::Daemon_reload] -> Service['daily.timer']
```
???

File: README.md (hunk 9 above)
Comment (alexjfisher) [reply README.md]: Yes, I think so.  This adds a `service` resource and the required (for puppet 5) relationship between daemon reload and the service.

File: manifests/timer.pp [hunk 12]
```diff
@@ -0,0 +1,107 @@
+# Create a timer and optionally a service unit
```
Comment (bastelfreak) [inline manifests/timer.pp]: this should be `# @summary Create a timer and optionally a service unit to execute with the timer unit`?

File: README.md (hunk 11 above)
Comment (traylenator) [reply README.md]: Seems bad to notify things inside the unit_file type , assumes it's contents.

Comment (alexjfisher) [reply README.md]: Not sure I follow. https://github.com/camptocamp/puppet-systemd/blob/f81339ad10516905e460d2899432ce92d83a35d2/manifests/unit_file.pp#L86 is the relationship I think you're duplicating.

Comment (alexjfisher) [reply README.md]: Actually, I think it's
```
Systemd::Timer['daily.timer'] ~> Service['daily.timer']
Class[Systemd::Systemctl::Daemon_reload] ~> Service['daily.timer']
```
that would be needed.

Perhaps systemd::timer can always make sure a service resource is declared itself? (which I think would be ok for timer units created with this module)
```puppet
unless $enable or $active {
  # if either $enable or $active are set, unit_file will create a service resource and 
  service { $name:
    provider  => systemd,
    subscribe => [
      Class[Systemd::Systemctl::Daemon_reload],
      Systemd::Unit_file[$name],
    ],
  }
}
```
or
```puppet
service { $name:
  provider  => systemd,
  enable    => $enable,
  active    => $active,
  subscribe => [
    Class[Systemd::Systemctl::Daemon_reload],
    Systemd::Unit_file[$name],
  ],
}
systemd::unit_file{$name:
    # as before but active/enable parameters removed
    ensure    => $ensure,
    content   => $timer_content,
    source    => $timer_source,
    path      => $path,
    owner     => $owner,
    group     => $group,
    mode      => $mode,
    show_diff => $show_diff,
}
```

File: README.md (hunk 9 above)
Comment (traylenator) [reply README.md]: Link added:

```
@see https://www.freedesktop.org/software/systemd/man/systemd.timer.html systemd.timer(5)
```
seems to be the correct syntax.

File: README.md (hunk 11 above)
Comment (traylenator) [reply README.md]: > Not sure I follow.
was me confused.

Creating the service inside timer.pp seems a shame to not reuse unit_file code.

File: README.md [hunk 13]
```diff
@@ -157,19 +164,14 @@ Type=oneshot
...
-
 systemd::timer{'daily.timer':
   timer_content   => $_timer,
   service_content => $_service,
 }
 
 service{'daily.timer':
   ensure => running,
-  require => Systemd::Timer['daily.timer'],
+  notify => Systemd::Timer['daily.timer'],
```
Comment (alexjfisher) [inline README.md]: ??? What does this do?

Comment (traylenator) [reply README.md]: I think it must do time travel.

Changed to subscribe - when the unit changes restart the service.

**Comment by traylenator** [comment]
rebased and squashed

File: manifests/timer.pp [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+#
+# @param timer_source
+#   The ``File`` resource compatible ``source``
+#
+#   * Mutually exclusive with ``$timer_content``
+#
+# @param service_content
+#   The full content of the service unit file
+#
+#   * Mutually exclusive with ``$timer_source``
```
Comment (ekohl) [inline manifests/timer.pp]:
```suggestion
#   * Mutually exclusive with ``$service_source``
```

File: manifests/timer.pp [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+) {
+  assert_type(Pattern['^.+\.timer$'],$name)
+
+  if $service_unit {
+    $_service_unit = $service_unit
+  } else {
+    $_service_unit = "${basename($name,'.timer')}.service"
+  }
+
+  if $service_content or $service_source {
```
Comment (ekohl) [inline manifests/timer.pp]: Within this block I'd expect something like:
```puppet
if $ensure == 'absent' {
  Systemd::Unit_file[$name] -> Systemd::Unit_file[$_service_unit]
} else {
  Systemd::Unit_file[$_service_unit] -> Systemd::Unit_file[$name]
}
```

That way you know the service is available when enabling the timer. If something goes wrong with defining the service, the timer will not be present. When removing you should go in the other direction.

File: README.md [hunk 16]
```diff
@@ -129,6 +129,82 @@ file { '/etc/tmpfiles.d/foo.conf':
 ~> Class['systemd::tmpfiles']
 ```
 
+### timer units
+Create a systemd timer unit and a systemd service unit to execute from
+that timer 
+
+The following will create a timer unit and a service unit file.
+The execution of `systemctl daemon-reload` will occur.
+As active and enable are set true the puppet service `runoften.timer` will be
```
Comment (ekohl) [inline README.md]:
```suggestion
When `active` and `enable` are set to `true` the puppet service `runoften.timer` will be 
```

File: README.md [hunk 17]
```diff
@@ -129,6 +129,82 @@ file { '/etc/tmpfiles.d/foo.conf':
...
 ```
 
+### timer units
+Create a systemd timer unit and a systemd service unit to execute from
+that timer 
+
+The following will create a timer unit and a service unit file.
+The execution of `systemctl daemon-reload` will occur.
+As active and enable are set true the puppet service `runoften.timer` will be 
+declared and be started and enabled.
```
Comment (ekohl) [inline README.md]:
```suggestion
declared, started and enabled.
```

**Comment by mmoll** [comment]
@traylenator please rebase :)

**Comment by mmoll** [comment]
ping?

**Comment by raphink** [comment]
@traylenator could you rebase your PR please?

**Comment by mmoll** [comment]
GH-138

**Comment by bastelfreak** [comment]
merged in https://github.com/camptocamp/puppet-systemd/pull/138

**Comment by traylenator** [comment]
Many thanks @mmoll
